dev/NavigationView/NavigationView.cpp

```diff
@@ -1592,6 +1770,161 @@ void NavigationView::UpdateVisualStateForDisplayModeGroup(const winrt::Navigatio
+    {
+        if (IsSettingsItem(nvi))
+        {
+            // Because ListViewItem eats the events, we only get these keys on KeyDown.
+            if (key == winrt::VirtualKey::Space ||
+                key == winrt::VirtualKey::Enter)
+            {
+                args.Handled(true);
+                OnSettingsInvoked();
+            }
+        }
+        else
+        {
+            switch (args.Key())
+            {
+            case winrt::VirtualKey::GamepadA:
+            case winrt::VirtualKey::Enter:
+            case winrt::VirtualKey::Space:
+                OnNavigationViewItemInvoked(nvi);
+                break;
+            case winrt::VirtualKey::Home:
+                KeyboardFocusFirstItemFromItem(nvi);
+                break;
+            case winrt::VirtualKey::End:
+                KeyboardFocusLastItemFromItem(nvi);
+                break;
+            }
+        }
+    }
+}
+
+void NavigationView::KeyboardFocusFirstItemFromItem(winrt::NavigationViewItemBase nvib)
+{
+    winrt::UIElement firstElement{ nullptr };
+    if (IsTopNavigationView())
+    {
+        bool isContainerInOverflow = IsContainerInOverflow(nvib);
+        if (isContainerInOverflow)
+        {
+            firstElement = m_topNavRepeaterOverflowView.get().TryGetElement(0);
+        }
+        else
+        {
+            firstElement = m_topNavRepeater.get().TryGetElement(0);
+        }
+    }
+    else
+    {
+        firstElement = m_leftNavRepeater.get().TryGetElement(0);
+    }
+
+    if (auto nvibFirst = firstElement.try_as<winrt::NavigationViewItemBase>())
+    {
+        nvibFirst.Focus(winrt::FocusState::Keyboard);
+    }
+}
+
+void NavigationView::KeyboardFocusLastItemFromItem(winrt::NavigationViewItemBase nvib)
+{
+    winrt::ItemsRepeater ir { nullptr };
+    if (IsTopNavigationView())
+    {
+        bool isContainerInOverflow = IsContainerInOverflow(nvib);
+        if (isContainerInOverflow)
+        {
+            ir = m_topNavRepeaterOverflowView.get();
+        }
+        else
+        {
+            ir = m_topNavRepeater.get();
+        }
+    }
+    else
+    {
+        ir = m_leftNavRepeater.get();
+    }
+
+    auto itemsSourceView = ir.ItemsSourceView();
+    if (itemsSourceView)
+    {
+        auto lastIndex = itemsSourceView.Count() - 1;
+        auto lastElement = ir.TryGetElement(lastIndex);
+        if (auto nvibLast = lastElement.try_as<winrt::NavigationViewItemBase>())
+        {
+            nvibLast.Focus(winrt::FocusState::Keyboard);
+        }
+    }
+}
+
+void NavigationView::OnNavigationViewItemKeyUp(const winrt::IInspectable& sender, const winrt::KeyRoutedEventArgs& args)
+{
+    if (!args.Handled())
+    {
+        if (auto nvi = sender.try_as<winrt::NavigationViewItem>())
+        {
+            if (IsSettingsItem(nvi))
+            {
+                // Because ListViewItem eats the events, we only get these keys on KeyUp.
+                if (args.OriginalKey() == winrt::VirtualKey::GamepadA)
```

Comment:
>OriginalKey [](start = 25, length = 11)

make this the first check, otherwise we will fire for every key and do multiple QIs and a comparison for no reason.